### PR TITLE
Fix mistake in sensitivity toolbox example and add unit test

### LINKED ISSUE #1535: Bug in sensitivity toolbox - not updating expression values
Here is a MWE adapted from ``parameter.py`` (example:
```
from pyomo.environ import * 
from pyomo.contrib.sensitivity_toolbox.sens import sipopt

def create_model():
    m = ConcreteModel()
    
    m.x1 = Var(initialize = 0.15, within=NonNegativeReals)
    m.x2 = Var(initialize = 0.15, within=NonNegativeReals)
    m.x3 = Var(initialize = 0.0, within=NonNegativeReals)
    
    m.eta1 = Param(initialize=4.5,mutable=True)
    m.eta2 = Param(initialize=1.0,mutable=True)
    
    m.const1 = Constraint(expr=6*m.x1+3*m.x2+2*m.x3-m.eta1 ==0)
    m.const2 = Constraint(expr=m.eta2*m.x1+m.x2-m.x3-1 ==0)
    m.cost = Objective(expr=m.x1**2+m.x2**2+m.x3**2)
    
    return m 


if __name__=='__main__':
    m = create_model()    

    m.perturbed_eta1 = Param(initialize = 4.0)
    m.perturbed_eta2 = Param(initialize = 1.0)


    m_sipopt = sipopt(m,[m.eta1,m.eta2],
                        [m.perturbed_eta1,m.perturbed_eta2],
                        streamSoln=True)
    
    print("\nOriginal parameter values:")
    print("\teta1 =",m.eta1())
    print("\teta2 =",m.eta2())
    
    print("Solution with the original parameter values:")
    print("\tObjective =",value(m.cost))
    print("\tx1 =",m.x1())
    print("\tx2 =",m.x2())
    print("\tx3 =",m.x3())
    
    print("\nNew parameter values:")
    print("\teta1 =",m_sipopt.perturbed_eta1())
    print("\teta2 =",m_sipopt.perturbed_eta2())
    
    print("(Approximate) solution with the new parameter values:")
    print("\tObjective =",value(m_sipopt.cost))
    print("\tx1 =",m_sipopt.x1())
    print("\tx2 =",m_sipopt.x2())
    print("\tx3 =",m_sipopt.x3())
```

The problem: The value stored in `m.cost` appears to be at the initial point, not the optimal solution.

### PR DESCRIPTION
**Pull request by adowling2**
## Summary/Motivation:
- Fixes `parameter.py` example to correctly access the initial point, nominal solution, and perturbed solution
- Fixes the same example in the OnlineDocs
- Adds unit test to cover `parameter.py` example
- Adds comments in `parameter.py` to explain a limitation of sipopt interface
- Adds `round()` to doctest to improve reliability
- Thank you @JanghoPark-LBL for pointing this out

## Changes proposed in this PR:
- [x] Modify `parameter.py` to highlight the bug
- [x] Test if the output model contains the base solution or the perturbed solution.
- [x] Update the example accordingly
- [x] Update the documentation accordingly
- [x] Add this example to the unit test library

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by adowling2** [comment]
@jsiirola This is an in-depth question about the function `sipopt` in `sens.py`. Here is the example `parameter.py`:

```
from pyomo.environ import * 
from pyomo.contrib.sensitivity_toolbox.sens import sipopt

def create_model():
    m = ConcreteModel()
    
    m.x1 = Var(initialize = 0.15, within=NonNegativeReals)
    m.x2 = Var(initialize = 0.15, within=NonNegativeReals)
    m.x3 = Var(initialize = 0.0, within=NonNegativeReals)
    
    m.eta1 = Param(initialize=4.5,mutable=True)
    m.eta2 = Param(initialize=1.0,mutable=True)
    
    m.const1 = Constraint(expr=6*m.x1+3*m.x2+2*m.x3-m.eta1 ==0)
    m.const2 = Constraint(expr=m.eta2*m.x1+m.x2-m.x3-1 ==0)
    m.cost = Objective(expr=m.x1**2+m.x2**2+m.x3**2)
    
    return m 


if __name__=='__main__':
    m = create_model()    

    m.perturbed_eta1 = Param(initialize = 4.0)
    m.perturbed_eta2 = Param(initialize = 1.0)


    m_sipopt = sipopt(m,[m.eta1,m.eta2],
                        [m.perturbed_eta1,m.perturbed_eta2],
                        streamSoln=True)
    
    print("\nOriginal parameter values:")
    print("\teta1 =",m.eta1())
    print("\teta2 =",m.eta2())
    
    print("Solution with the original parameter values:")
    print("\tObjective =",value(m.cost))
    print("\tx1 =",m.x1())
    print("\tx2 =",m.x2())
    print("\tx3 =",m.x3())
    
    print("\nNew parameter values:")
    print("\teta1 =",m_sipopt.perturbed_eta1())
    print("\teta2 =",m_sipopt.perturbed_eta2())
    
    print("(Approximate) solution with the new parameter values:")
    print("\tObjective =",value(m_sipopt.cost))
    print("\tx1 =",m_sipopt.x1())
    print("\tx2 =",m_sipopt.x2())
    print("\tx3 =",m_sipopt.x3())
```

By default, `sipopt` creates a copy (`.clone()`) for the original Pyomo model. Interestingly, the optimal solution, before the perturbation, is accessible in the original model `m`. But, the objective is incorrect. I think the expression `m.cost` is not being updated.

**Comment by adowling2** [comment]
For completeness, here is the output I get from `paramester.py` (on this PR/branch):
```
Ipopt 3.13.2: run_sens=yes


******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt
******************************************************************************

This is Ipopt version 3.13.2, running with linear solver ma27.

Number of nonzeros in equality constraint Jacobian...:       10
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:        4

Total number of variables............................:        5
                     variables with only lower bounds:        3
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equality constraints.................:        4
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  4.5100000e-02 4.50e+00 4.07e-01  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  8.9566717e-01 1.32e-01 6.64e+00  -1.0 4.50e+00    -  1.85e-01 1.00e+00h  1
   2  5.8493209e-01 8.88e-16 5.62e-01  -1.0 2.59e-01    -  9.90e-01 1.00e+00f  1
   3  5.2819977e-01 8.88e-16 1.29e-02  -1.7 8.27e-02    -  1.00e+00 1.00e+00f  1
   4  5.0741726e-01 8.88e-16 4.99e-03  -2.5 4.19e-02    -  1.00e+00 1.00e+00f  1
   5  5.0084750e-01 8.88e-16 7.29e-04  -3.8 1.62e-02    -  1.00e+00 1.00e+00f  1
   6  5.0001574e-01 8.88e-16 1.47e-05  -5.7 2.20e-03    -  1.00e+00 1.00e+00f  1
   7  5.0000000e-01 8.88e-16 5.54e-09  -8.6 4.20e-05    -  1.00e+00 1.00e+00f  1

Number of Iterations....: 7

                                   (scaled)                 (unscaled)
Objective...............:   4.9999999772520909e-01    4.9999999772520909e-01
Dual infeasibility......:   5.5407590737743817e-09    5.5407590737743817e-09
Constraint violation....:   8.8817841970012523e-16    8.8817841970012523e-16
Complementarity.........:   7.7252085982569584e-09    7.7252085982569584e-09
Overall NLP error.......:   7.7252085982569584e-09    7.7252085982569584e-09


Number of objective function evaluations             = 8
Number of objective gradient evaluations             = 8
Number of equality constraint evaluations            = 8
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 8
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 7
Total CPU secs in IPOPT (w/o function evaluations)   =      0.002
Total CPU secs in NLP function evaluations           =      0.000

EXIT: Optimal Solution Found.

Original parameter values:
	eta1 = 4.5
	eta2 = 1.0
Solution with the original parameter values:
	Objective = 0.045
	x1 = 0.15
	x2 = 0.15
	x3 = 0.0

New parameter values:
	eta1 = 4.0
	eta2 = 1.0
(Approximate) solution with the new parameter values:
	Objective = 0.4999999977252091
	x1 = 0.5000000037913185
	x2 = 0.4999999939338906
	x3 = 0.0
```

And this answers my question. Pyomo model `m` only contains the initial point, not the solution.

**Comment by adowling2** [comment]
I verified the new doctest works on my local machine.

**Comment by adowling2** [comment]
@jsiirola @JanghoPark-LBL This is ready to review.

**Comment by adowling2** [comment]
@JanghoPark-LBL Thank you! I updated the documentation.

File: pyomo/contrib/sensitivity_toolbox/tests/test_sens.py [hunk 1]
```diff
@@ -351,6 +351,22 @@ def test_constraintSub(self):
                        len(list(identify_variables(
                                 m_sipopt._sipopt_data.constList[2].body))) == 2)
 
+    # Test example `parameter.py`
+    @unittest.skipIf(not opt.available(False), "ipopt_sens is not available")
+    def test_parameter_example(self):
+    
+        from pyomo.contrib.sensitivity_toolbox.examples.parameter import example
+        d = example(print_flag = False)
```
Comment (jsiirola) [inline pyomo/contrib/sensitivity_toolbox/tests/test_sens.py:359]: Why turn off the print flag when testing the example?

**Pull request merged**